rust: Use a trait-based client API

Zakru opened the pull request:
This replaces the API with one based around a `Client` trait.

Benefits:
- Idiomatic API
- `static`s are managed by the API implementation
- `extern`s are declared in a controlled way
- Minimal abstraction overhead
  - Theoretically zero-cost; the client functions should get inlined into the exported functions (with I/O parsing/encoding boilerplate), but I have not done any decompilation to check.
  - This is made possible by some well-documented `unsafe` code.

Minimal example:

```rust
use wasmbot_client::{Client, params, register_client, wasmbot_messages};

struct Bot;

impl Client for Bot {
    fn create() -> Self {
        Self
    }

    fn get_metadata(&mut self) -> params::BotMetadata {
        params::BotMetadata {
            name: params::make_bot_name("test bot"),
            version: [1, 0, 0],
        }
    }

    fn tick(&mut self, _pc: wasmbot_messages::PresentCircumstances) -> wasmbot_messages::Message {
        wasmbot_messages::Message::Wait(wasmbot_messages::Wait {  })
    }
}

register_client!(Bot);
```

Inspired by: https://github.com/zetanumbers/wasm4-rs/blob/main/wasm4/src/rt.rs

sjml commented:
Thanks for this! I'm doing some work in the next branch for a little bit but will check this out soon!

sjml commented:
Before I look more, though, could you change it to use tabs instead of spaces? (My overall policy as I was doing this was to use tabs unless it was absolutely mandated by the programming language or its first party fmt. I'm not overzealous about that, so willing to revisit, but in the meantime, it makes the diffs much simpler if we keep whitespace formatting.)

Zakru commented:
Marking as draft for a few reasons.

First, addressing your comment, apologies for not conforming to the repo's existing style. However, Rust's default style guide (https://doc.rust-lang.org/nightly/style-guide/index.html) recommends 4 spaces for indentation, and indeed the `rustfmt` formatter by default enforces this, so my personal suggestion would be to switch to that. I did notice that `cargo fmt` removes left-padding for positive numbers in `explorer.rs`, so I didn't want to run it yet as-is.

Secondly, I was only checking Wasm builds and neglected the "trainer" portion, and I introduced a compile error to the client library's native build. I also can't get `make trainer` to build out of the box on Windows (even in the `main` branch), as the current build script doesn't find the compiled bot lib:

```
LINK : fatal error LNK1181: cannot open input file 'circler.lib'
```

sjml commented:
> First, addressing your comment, apologies for not conforming to the repo's existing style. 

No apology necessary! I sincerely appreciate the efforts you're putting in here! :D 

> However, Rust's default style guide (https://doc.rust-lang.org/nightly/style-guide/index.html) recommends 4 spaces for indentation, and indeed the `rustfmt` formatter by default enforces this, so my personal suggestion would be to switch to that. 

I knew the recommendation, but it's not an absolute requirement like tabs in Go or (to a somewhat lesser extent) spaces in Zig, so at the time I went with tabs. I must have forgotten to run `cargo fmt` since I made the switch. 😳 (These are the things that can slip your mind when you have six programming languages + shell scripts in a single repo.) 

Let's leave as tabs for the sake of this PR and I'll look at switching back to spaces separately. (Would want to change over the validator, too.)

> <snip problems with Windows build>

I sadly don't have a Windows machine to test on at the moment, and won't until I return to the US at the end of the summer. I would very much like to have this all working on Windows, but can't do the fixes myself at the moment. The trainer is a secondary priority to the wasm, but I wouldn't want to break it unnecessarily. When I look at this more fully I'll at least get it working on my Mac, and then can leave the problem of Windows trainers to some future iteration of myself. (Unless of course you feel like tracking down the build issue!)

So I'd say don't sweat it too much, but if it's an easy and obvious fix, feel free to apply here.

Zakru commented:
I fixed up the style and revamped how the trainer is built. I can now successfully build and run the trainer bin in both WSL (with `make` or just Cargo) and native Windows. I can't get the `deno-cli` trainer program to work on Windows probably due to some path magic issue, but it seems to work fine in WSL.

<details>
<summary>Deno error</summary>

```
error: Uncaught (in promise) NotFound: The system cannot find the path specified. (os error 3): readfile '/C:/Users/Zakru/git/wasmbots/engine/rsc/maps/tilesets/dungeon_tiles.tsj'
                        return await Deno.readTextFile(filepath);
                                          ^
    at Object.readTextFile (ext:deno_fs/30_fs.js:784:24)
    at readTextFile (file:///C:/Users/Zakru/git/wasmbots/engine/src/core/loader.ts:11:22)
    at readJsonFile (file:///C:/Users/Zakru/git/wasmbots/engine/src/core/loader.ts:25:21)
    at Function.loadDynamic (file:///C:/Users/Zakru/git/wasmbots/engine/src/game/map.ts:91:25)
    at World.setMap (file:///C:/Users/Zakru/git/wasmbots/engine/src/game/world.ts:140:29)
    at main (file:///C:/Users/Zakru/git/wasmbots/engine/src/deno-cli/trainer.ts:29:14)
    at eventLoopTick (ext:core/01_core.js:175:7)
```
</details>

The trainer program is now built as a Cargo example in the bot project instead of manually linking across crates (discouraged). I turned the `trainer` from a bin crate to a lib crate, with its main function now being called from the bot's example.

The updated Makefile works 1:1.

I think that should resolve the remainder of the issues mentioned in the thread.

sjml commented:
Thanks, I'll take a look soon! (Not surprised there's issues at that level with Windows; I'm guessing the symlink stuff which is already a little brittle on macOS just falls apart there. Whole thing could use a restructuring at some point. In any case, don't worry about it on your end!)

abdavis commented:
If you _really_ want to use tabs instead of spaces, consider adding a `rustfmt.toml` file with `hard_tabs = true`. Many rust programmers have setups that auto format on save, this will save some headaches. Honestly I like to think as little as possible about formatting and I just go with the defaults. In the grand scheme of things tabs vs spaces matters little to me, so why go against the grain?

sjml commented:
> Honestly I like to think as little as possible about formatting and I just go with the defaults. In the grand scheme of things tabs vs spaces matters little to me, so why go against the grain?

On the whole, I agree, but I was pointed to this post by a few programmers I really respect and found it persuasive. You may not, and that is ok! https://www.reddit.com/r/javascript/comments/c8drjo/nobody_talks_about_the_real_reason_to_use_tabs/

abdavis commented:
Fair enough, you get to decide! The main point I wanted to get across is you can change the default behavior.

sjml commented:
It's seeming more and more likely that I'll break this monorepo apart (since people actually seem interested in the project, which is cool!) and at that point I'll be more tolerant of inconsistencies. :)

sjml commented on example_bots_src/circler/Cargo.toml:
```diff
@@ -11,5 +11,10 @@ strip = true
+[dev-dependencies]
+trainer = { path = "../../libraries/Rust/trainer" }
+
+[[example]]
```
Not sure where else to put this question, so just putting it here -- is "example" / "examples" an idiomatic thing? Since it's already in an examples directory, it just feels a little redundant. Unless this is just a convention I'm not aware of...

Zakru replied:
Examples are a built-in Cargo feature, and by default they are placed in an `examples` dir. Among other conveniences, they get built along with `cargo test` to make sure they don't break (which should probably be run in a GH action). The explicit declaration here is redundant, though, I removed it.

https://doc.rust-lang.org/cargo/reference/cargo-targets.html#examples

sjml replied:
Makes sense... I'm probably overthinking, but the trainer doesn't really feel like an example. Like a dev binary or something? If examples is the best fit, so be it. I'm probably thinking too much.

sjml replied:
And yeah, would love to add some automated testing, but things are pretty fluid at the moment. The ci that does the deployment to pages is effectively a very minimal test, since it build all the bots and the frontend.

Zakru replied:
This SO issue has a nice list of options for a similar case, considering the trainer crate is a bin-only dependency https://stackoverflow.com/a/46017284/3194479
Basically
- Have the binary in the crate itself
  - A bit of hassle with crate features, need to add a `--features ...` flag when building
- Example
- Separate crate
  - A Cargo workspace would be used to keep them as one project

This is structurally kind of an outlier case imo. I think if the bot was compiled to a native shared library and driven by a program implemented only once (instead of the `libraries/*/trainer` program for each language) that would be the most accurate setup. Don't know how viable that is, though.

sjml replied:
Yeah, that's more-or-less the journey I went through when I originally set this up. I recognize it's a weird model here, but I wish that Rust let me define a main function in a crate that gets included as a dependency. This is like the most edgiest of edge cases, but for what's going on here, it's actually kinda what we want? We're sort of stuck building the bot as a library because of how the wasm support works in Rust. You are right, though, that this is basically the edgiest of edge cases. 

I don't think it's viable to have the one trainer program, at least not within my scope. That would be ideal, but given the different ABIs and linking conventions and everything, I think each trainer just kinda has to be its own thing.

sjml commented on example_bots_src/circler/Makefile:
```diff
@@ -1,14 +1,14 @@
-	BOT_LIBRARY_NAME=$(EXE_NAME) BOT_LIBRARY_PATH=../../../example_bots_src/circler/target/debug/ cargo build --manifest-path ../../libraries/Rust/trainer/Cargo.toml --target-dir ./target
-	ln -s ./target/debug/trainer ./$(EXE_NAME)-trainer
+trainer: $(BOT_SRCS) $(TRAINER_SRCS)
+	cargo build --example trainer
```
Ah, ok I see what you're doing. I like it... just going 🤨 at the repetition but I'm almost certainly overthinking this.

Zakru replied:
Technically this could be in a separate binary crate (in a single Cargo workspace so it's still effectively the same project), but I feel like that's a bit overkill for 4 LoC

sjml replied:
Yeah, agreed; if/when this gets broken out into a separate repository we can think more about overall structure.

sjml commented on libraries/Rust/trainer/src/lib.rs:
```diff
@@ -29,9 +29,18 @@ struct MemoryResult {
+/// Runs the trainer.
+/// 
+/// The client must be linked to the program.
+pub fn run() {
```
Is this single-threaded? Probably not too important but just to remove one more axis of bugs. (If it is, all good!)

Zakru replied:
Yes, this is effectively `#[tokio::main(flavor = "current_thread")]` expanded.

https://docs.rs/tokio/latest/tokio/attr.main.html#using-current-thread-runtime

sjml reviewed (changes requested):
This is excellent work, and I really appreciate it. I threw a few questions in, but no rush. Basically ready to go once those are answered. :)

Zakru commented:
Fixed an editing mistake that I managed to create after running my tests last time, while I was at it
